### PR TITLE
Improve tokenizer aw

### PR DESCRIPTION
**Pull request by zredlined**
* added progress bar for sentencepiece tokenizer (can take a while on large datasets)
* cleaned up logging
* corrected invalid calculation of model perplexity
* export model params and training history to model directory
* config option to save best, or all checkpoints (save disk space)

### CODE REVIEW

File: src/gretel_synthetics/train.py [hunk 1]
```diff
@@ -24,41 +28,71 @@
...
+
+
+def save_model_params(store: BaseConfig):
+    save_path = Path(store.checkpoint_dir) / 'model_params.json'
+    logging.info(f"Saving model history to {save_path.name}")
+    with open(save_path, 'w') as f:
+        json.dump(store.as_dict(), f, indent=2)
+
+
+def save_history_csv(history:LossHistory, store:BaseConfig):
```
Comment (johntmyers) [inline src/gretel_synthetics/train.py:49]: whitespace between param and typehint

File: src/gretel_synthetics/config.py [hunk 2]
```diff
@@ -64,3 +65,13 @@ def _set_tokenizer(self):
...
+
+    def as_dict(self):
+        config_dict = {key: value for key, value in self.__dict__.items()
+                       if not key.startswith('_') and not callable(key)}
+
+        # delete temporary configuration items
+        for key in ['tokenizer_prefix', 'tokenizer_model', 'training_data', 'tokenizer', 'processed_data']:
+            del config_dict[key]
+
+        return config_dict
```
Comment (johntmyers) [inline src/gretel_synthetics/config.py:77]: We should probably look to make the config base class a dataclass. This will give us more natural serialization. I can take a look at this if you like

File: src/gretel_synthetics/train.py (hunk 1 above)
Comment (johntmyers) [inline src/gretel_synthetics/train.py:49]: Should these be methods on the config class so they can be called directly from the store instance?

File: src/gretel_synthetics/train.py [hunk 3]
```diff
@@ -24,41 +28,71 @@
...
 
-    all_cbs = [checkpoint_callback]
-
-    logging.info("Training model")
-    model.fit(dataset, epochs=store.epochs, callbacks=all_cbs)
-    logging.info(
-        f"Wrote latest checkpoint to disk: {tf.train.latest_checkpoint(store.checkpoint_dir)}")
+    model.fit(dataset, epochs=store.epochs, callbacks=[checkpoint_callback, history_callback])
+    save_history_csv(history_callback, store)
+    save_model_params(store)
```
Comment (johntmyers) [inline src/gretel_synthetics/train.py:86]: See other comment, should these be something like `store.save_model_params()` and `store.save_history_csv(history_callback)` ?

File: src/gretel_synthetics/train.py [hunk 4]
```diff
@@ -121,8 +154,13 @@ def create_dataset(store: BaseConfig, text: str, sp: spm.SentencePieceProcessor)
     Create two lookup tables: one mapping characters to numbers,
     and another for numbers to characters.
     """
-    # Create training dataset
-    char_dataset = tf.data.Dataset.from_tensor_slices(sp.EncodeAsIds(text))
+    logging.info("Tokenizing training data")
+    ids = []
+    for line in tqdm(text.split("\n")):
+        ids = ids + sp.EncodeAsIds(line)
```
Comment (johntmyers) [inline src/gretel_synthetics/train.py:160]: can we `ids.append(sp.EncodeAsIds(line)` here?

File: src/gretel_synthetics/model.py [hunk 5]
```diff
@@ -61,13 +53,11 @@ def build_sequential_model(vocab_size: int, batch_size: int, store: BaseConfig)
                                                              reduction=tf.losses.Reduction.NONE)
 
     else:
-        logging.info("Utilizing non-private RMSProp optimizer")
+        logging.warning("Differentially private training _not_ enabled")
```
[the remark this answers is not in the document]
Comment (zredlined) [reply src/gretel_synthetics/model.py:56]: just in case anyone has markdown enabled in their terminal =)

File: src/gretel_synthetics/train.py (hunk 1 above)
Comment (zredlined) [reply src/gretel_synthetics/train.py:49]: 1. For save model params, yes. Will refactor.
2. For saving the model training history (loss/accuracy), I think it makes sense to keep it in train.py

File: src/gretel_synthetics/train.py (hunk 4 above)
Comment (zredlined) [reply src/gretel_synthetics/train.py:160]: Need to keep as one contiguous array, but can simplify to `ids += sp.EncodeAsIds(line)`

Comment (johntmyers) [reply src/gretel_synthetics/train.py]: Ah right. Then use extend()

**Comment by zredlined** [comment]
Moving updates to config class to a new ticket (data class, tests). Merging.
